state/charm.go

```diff
@@ -528,7 +529,43 @@ func (c *Charm) Meta() *charm.Meta {
 
 // Config returns the configuration of the charm.
 func (c *Charm) Config() *charm.Config {
-	return c.doc.Config
+	model, err := c.st.Model()
```

Comment:
This isn't what you want to do.

What you need to do is have the Update() method on the application facade split the config into app and charm config the same as is done for SetApplicationsConfig() on that facade or func deployApplication() in apiserver/facades/client/application/application.go